### PR TITLE
Documenting review and consolidation

### LINKED ISSUE #88: (On going) Review Process
An outline of review schema and process, for discussion and reference (to be turned into internal docs and broken down into an epic)

### PR DESCRIPTION
**Pull request by andreievg**
Fixes #88 

Above issue started explain consolidation and review process but got outdated. Most of the goals of the above issue were accomplished by the design, diagrams and discussions over the last month (only some aspects of assigner role remain, there is an issue for that https://github.com/openmsupply/application-manager-web-app/issues/294)

Explains most of the review and consolidation implementation. Some of the front end logic is also explained here, linked through this PR: https://github.com/openmsupply/application-manager-web-app/pull/309

Can't quite see in the diff, but I did try to keep most of the existing review docs

### CODE REVIEW

File: documentation/Database-Schema-Review-And-Consolidations.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
+# Database Schema Review
+
+Original thinking was for review and consolidation to be separate entities, upon looking at intersection in schemas, it was apperant that they should be combined. Consolidation is really the review of a review vs review of application.
```
Comment (CarlosNZ) [inline documentation/Database-Schema-Review-And-Consolidations.md:3]:
```suggestion
Original thinking was for review and consolidation to be separate entities, upon looking at intersection in schemas, it was apparent that they should be combined. Consolidation is really the review of a review vs review of application.
```

File: documentation/Database-Schema-Review-And-Consolidations.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+## Review Decision
+
+Only available on level > 1 reviewer or when `is_last_level` in `review_assignment` is set. Consolidators can make/suggest decisions. If there is only one review level (say for screening), then the screener has the ability to create review decisions
+
+`Conform (Approve)` -> All of the applicant's responses must be marked as approved by reviewer, and consolidator agreed with reviewers decision
+
+`LOQ` -> At least one of applicant's responses marked as non conformity (Decline) and consolidator agreed with reviewers decisions. They have an option to select from LOQ or Non-Conformity
+
+`Non-conofrm (Decline` -> At least one of applicant's responses marked as non conformity (Decline) and consolidator agreed with reviewers decisions. They have an option to select from LOQ or Non-Conformity
```
Comment (CarlosNZ) [inline documentation/Database-Schema-Review-And-Consolidations.md:19]:
```suggestion
`Non-conform (Decline)` -> At least one of applicant's responses marked as non conformity (Decline) and consolidator agreed with reviewers decisions. They have an option to select from LOQ or Non-Conformity
```

File: documentation/Database-Schema-Review-And-Consolidations.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+- Determine if review can be started/created <- front end can check if review can be created (including self assignment) See `Review and Consolidation URL flow` diagram below
+
+#### Review Assignment Status
+
+`Available` -> can be assigned by assigner
+`Not Available` -> cannot be assigned by assigner, in case of `available for self assignment`, when reviewer self assigns a review, all other `review assignments` for current `stage` and `level` are changed to this status
+`Assigned` -> review is assigned to a user and can be started (in case of self assignment when review is self assigned, review assignment will changes status to assigned)
+`Available for self assignment` -> review can be started and self assigned
+
+`Assignment NOTE` -> For MVP, only level 1 reviewer can be assigned questions (vs self assignment of review), level > 1 will be self assigned (level 1 can also be configured to self assigned, in this case they will need to be configured to review all sections)
```
Comment (CarlosNZ) [inline documentation/Database-Schema-Review-And-Consolidations.md:39]: These need either double-spaces at the end of each to force a line-break, or space between to make separate paragraphs. Currently they render all in one continuous paragraph.

File: documentation/Database-Schema-Review-And-Consolidations.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+##### Other fields (that are not straight away self explanatory)
+
+`assigner_id` -> null until assigned, assigner id when assigned be assigner, otherwise reviewer_id when self assigned
+`is_last_level` -> last level for current stage (determine if review_decision is present in UI)
+`allowable_section_ids` -> an array of section IDs that reviewer has permission to review (would typically have all sections), assigner can only assign questions from sections that are in this list
```
Comment (CarlosNZ) [inline documentation/Database-Schema-Review-And-Consolidations.md:45]: Same for these.

File: documentation/Database-Schema-Review-And-Consolidations.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+## Review Question Assignment
+
+Identifies which questions, from the application can be reviewed by the reviewer.
+These records are created upon assignment, and should only be created for level 1 reviewer ? (to confirm)
```
Comment (CarlosNZ) [inline documentation/Database-Schema-Review-And-Consolidations.md:50]: Only for level 1? I guess that could work, just wondering which has less convoluted logic. Did we discuss this yet?

File: documentation/Database-Schema-Review-And-Consolidations.md [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+## Review Response
+
+Either a review of applications questions (base review, level 1) or review of review (consolidation, level > 1). See diagram (responses flow). Always created by front end, but duplicates are trimmed on back end.
+Can be used to agree or disagree on overall review_decision.
+
+A few trick bits:
+
+##### Review Responses Are Always Duplicated i
```
Comment (CarlosNZ) [inline documentation/Database-Schema-Review-And-Consolidations.md:59]:
```suggestion
#### Review Responses Are Always Duplicated
```

File: documentation/Database-Schema-Review-And-Consolidations.md [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+During re-review, or in reply to changes requested, all existing review responses are duplicated (so that any changes are recorded in a new record). They are trimmed by an action on back end.
+
+##### Review Submitted When Consolidation is in Progress
```
Comment (CarlosNZ) [inline documentation/Database-Schema-Review-And-Consolidations.md:63]:
```suggestion
#### Review Submitted When Consolidation is in Progress
```

File: documentation/Database-Schema-Review-And-Consolidations.md [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+### Creating and editing review
+
+Whenever user starts a review (either first time or subsequent times that they can edit it after submissions), we would create all responses, or duplicate them from existing responses. Trimming is done for unchanged responses after submission via an action (above rules should take into account duplicates). Slight difference for consolidation is we only duplicate review_response that are not 'draft' and for reviews that are submitted
+
+### Diagram
+
+![Review Life Cycle](images/review-life-cycle.png)
+![Response Flow](images/response-flow.png)
+![Review and Consolidation URL flow](images/consolidation-and-review-URL-flow)
```
Comment (CarlosNZ) [inline documentation/Database-Schema-Review-And-Consolidations.md:116]: Missing image file extension (.png?)

**Review by CarlosNZ — CHANGES_REQUESTED** [review]
Few little typos and formatting issues. 👍

File: documentation/Database-Schema-Review-And-Consolidations.md (hunk 3 above)
Comment (andreievg) [reply documentation/Database-Schema-Review-And-Consolidations.md]: Oops, I should have been more diligent in checking formatting:

85e16cb2a3d5923c19c556128122fd2fcc9a2ece

File: documentation/Database-Schema-Review-And-Consolidations.md (hunk 4 above)
Comment (andreievg) [reply documentation/Database-Schema-Review-And-Consolidations.md]: 85e16cb2a3d5923c19c556128122fd2fcc9a2ece

File: documentation/Database-Schema-Review-And-Consolidations.md (hunk 5 above)
Comment (andreievg) [reply documentation/Database-Schema-Review-And-Consolidations.md:55]: In the diagrams that were created, this is taken into account (i am not sure if it's demonstrated). 

i.e. level one reviewer responses are created based on review question assignments but lvl +1 review responses are created based on submitted reviews review responses from lvl -1

let me know if you want to chat about this

File: documentation/Database-Schema-Review-And-Consolidations.md (hunk 8 above)
Comment (andreievg) [reply documentation/Database-Schema-Review-And-Consolidations.md]: 34c53ba5c4be4d73b727632baf113bbf090f30c4

**Comment by andreievg** [comment]
@CarlosNZ thanks for the review ! I've made the changes and committed suggestions

File: documentation/Database-Schema-Review-And-Consolidations.md [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+`Available` -> can be assigned by assigner
+
+`Not Available` -> cannot be assigned by assigner, in case of `available for self assignment`, when reviewer self assigns a review, all other `review assignments` for current `stage` and `level` are changed to this status
+
+`Assigned` -> review is assigned to a user and can be started (in case of self assignment when review is self assigned, review assignment will changes status to assigned)
+
+`Available for self assignment` -> review can be started and self assigned
+
+`Assignment NOTE` -> For MVP, only level 1 reviewer can be assigned questions (vs self assignment of review), level > 1 will be self assigned (level 1 can also be configured to self assigned, in this case they will need to be configured to review all sections)
```
Comment (nmadruga) [inline documentation/Database-Schema-Review-And-Consolidations.md:42]: Hmm this bit is a little unclear to me. Is this a different type of status?

**Review by CarlosNZ — APPROVED** [review]
LGTM!

File: documentation/Database-Schema-Review-And-Consolidations.md [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+##### Other fields (that are not straight away self explanatory)
+
+`assigner_id` -> null until assigned, assigner id when assigned be assigner, otherwise reviewer_id when self assigned
+
+`is_last_level` -> last level for current stage (determine if review_decision is present in UI)
+
+`allowable_section_ids` -> an array of section IDs that reviewer has permission to review (would typically have all sections), assigner can only assign questions from sections that are in this list
```
Comment (nmadruga) [inline documentation/Database-Schema-Review-And-Consolidations.md:50]: At some point we agreed that when no sections were specified we would say the user can be assigned to any.
Are we changing to have all listed at all non-restricted `review_assignments`?

File: documentation/Database-Schema-Review-And-Consolidations.md (hunk 5 above)
Comment (nmadruga) [reply documentation/Database-Schema-Review-And-Consolidations.md:55]: Yea, if we are not creating for other levels we need to be aware that the front-end need to be changed to create the associated review_responses differently. For now it's using the existing `review_question_assignment` to create the `review_responses` here:

https://github.com/openmsupply/application-manager-web-app/blob/d49b5e9fedad4c5b1a7c94aa0a5db63d520dc80b/src/containers/Review/ReviewOverview.tsx#L50

File: documentation/Database-Schema-Review-And-Consolidations.md [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+
+## Review Question Assignment
+
+Identifies which questions, from the application can be reviewed by the reviewer.
+These records are created upon assignment, and should only be created for level 1 reviewer ? (to confirm)
+
+## Review Response
+
+Either a review of applications questions (base review, level 1) or review of review (consolidation, level > 1). See diagram (responses flow). Always created by front end, but duplicates are trimmed on back end.
```
Comment (nmadruga) [inline documentation/Database-Schema-Review-And-Consolidations.md:59]:
```suggestion
Either a review of applications questions (base review, level 1) or review of reviews (consolidation, level > 1). See diagram (responses flow). Always created by front end, but duplicates are trimmed on back end.
```

File: documentation/Database-Schema-Review-And-Consolidations.md [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+Can be used to agree or disagree on overall review_decision.
```
Comment (nmadruga) [inline documentation/Database-Schema-Review-And-Consolidations.md:60]: Is this related to the future field (in review_response) `review_decision_id`?

**Review by nmadruga — COMMENTED** [review]
Very good to have this one in @andreievg!

I can see that this has been already merged now, I just have 2 things to add here:

On the `database-schema-review-and-consolidation` ER diagram the field in red `review_decision_id` is like that since it isn't included in the schema yet - to be done in another version when we support the consolidator to overwrite some lower level decision. I think it would be good to comment that on the documentation somewhere.

On the `consolidation-and-review-URL-flow` flow diagram the **Review Statuses** are outdated with the ones in use, I wonder if we should: 
a. Change in the flow digram to be consistent with current ones in schema; OR
b. Update schema to use `pending_review` instead of `pending` (and update other documents with it). I think I like this name better than just pending.

**Pull request merged**